### ISSUE TITLE
Unable to compress messages

### ISSUE THREAD

**Issue by dkoston** [segmentio/kafka-go#117]
This is likely user error but I'm not able to compress messages.

I've set them to use `snappy` with the exact code from README.md (have also tried lz4):

```
	w := kafka.NewWriter(kafka.WriterConfig{
		Brokers: []string{"localhost:9092"},
		Topic:   "topic-A",
		CompressionCodec: snappy.NewCompressionCodec(),
	})
```

However, when the messages reach Kafka, they are not compressed:

![screen shot 2018-09-14 at 4 02 59 pm](https://user-images.githubusercontent.com/6424250/45575043-b38eea80-b837-11e8-8e09-87d339c10cf6.png)

Code: https://gist.github.com/dkoston/aab79446dc411c9d125a412ea20802c4

Perhaps there's some broker configuration I need to do? I was under the impression that each message itself contains bits telling the broker what compression codec is used on it's contents and it's unnecessary to specify a codec at the broker level if you do so at the producer/writer.

If I add some debug to the client, I see that the messages are running through `transform()` so I'd expect them to be compressed with lz4 and marked as such:

```
$ go run main.go
2018/09/14 15:41:29 Using snappy for message compression
2018/09/14 15:41:29 Writing 10 messages to topic-A
2018/09/14 15:41:30 [kafka-go]: compressing
2018/09/14 15:41:30 [kafka-go]: compressing
2018/09/14 15:41:30 [kafka-go]: compressing
2018/09/14 15:41:30 [kafka-go]: compressing
2018/09/14 15:41:30 [kafka-go]: compressing
2018/09/14 15:41:30 [kafka-go]: compressing
2018/09/14 15:41:30 [kafka-go]: compressing
2018/09/14 15:41:30 [kafka-go]: compressing
2018/09/14 15:41:30 [kafka-go]: compressing
2018/09/14 15:41:30 [kafka-go]: compressing
2018/09/14 15:41:30 Reading 10 messages from topic-A as group1
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 15:41:42 [kafka-go]: not compressed: 0
2018/09/14 16:03:36 |----------|--------------|---------------|
2018/09/14 16:03:36 | Messages | Uncompressed | Compressed    |
2018/09/14 16:03:36 |----------|--------------|---------------|
2018/09/14 16:03:36 |       10 |           10 |             0 |
2018/09/14 16:03:36 |----------|--------------|---------------|
```

I added:

```
diff --git a/message.go b/message.go
index 2d8b775..12ed4c2 100644
--- a/message.go
+++ b/message.go
@@ -5,6 +5,7 @@ import (
        "bytes"
        "fmt"
        "time"
+       "log"
 )

 // Message is a data structure representing kafka messages.
@@ -54,9 +55,12 @@ func (msg Message) message() message {

 func (msg Message) encode() (Message, error) {
        if msg.CompressionCodec == nil {
+               log.Printf("[kafka-go]: not compressing")
                return msg, nil
        }

+       log.Printf("[kafka-go]: compressing")
+
        var err error
        msg.Value, err = transform(msg.Value, msg.CompressionCodec.Encode)
        return msg, err
@@ -65,6 +69,7 @@ func (msg Message) encode() (Message, error) {
 func (msg Message) decode() (Message, error) {
        c := msg.message().Attributes & compressionCodecMask
        if c == CompressionNoneCode {
+               log.Printf("[kafka-go]: not compressed: %v", c)
                return msg, nil
        }

@@ -73,6 +78,8 @@ func (msg Message) decode() (Message, error) {
                return msg, fmt.Errorf("codec %d not imported.", msg.CompressionCodec)
        }

+       log.Printf("[kafka-go]: compressed")
+
        var err error
        msg.Value, err = transform(msg.Value, codec.Decode)
        return msg, err
```

**Comment by dkoston** [comment]
note: I've tried with kafka 2.0.0 and 1.1.0. It's not clear what versions are supported by this library

**Comment by abraithwaite** [comment]
Hi @dkoston !

We've been working on API version support for a while now (it's pretty hairy with how the kafka wire protocol is defined): https://github.com/segmentio/kafka-go/pull/80

To answer your question though, IIRC the reader automatically decompresses the message.  Verifying that now though.

**Comment by Pryz** [comment]
Hey @dkoston, @abraithwaite is right, the consumer (reader) will uncompress the payload automatically. If you take a look at https://github.com/segmentio/kafka-go/blob/master/message.go#L66 you will see that the reader gets the compression codec from the message attributes and decode the message if needed.

For the supported Kafka versions : due to how Kafka manages message formats, kafka-go will work with any Kafka version but you will only be able to use features available in Kafka 0.10 (and 0.11 for some of them like consumer groups). We are currently rethinking the internal of the library to handle more features without introducing API changes.

**Comment by dkoston** [comment]
Hi @abraithwaite! Thanks for the library btw.

Definitely understand the challenges w/ kafka wire protocol and versions.

I'm aware that the reader does auto decompression on the messages. What I'm looking to do on the reader side is determine what type of compression was used on each message for other purposes.

The code is accessing `msg.CompressionCodec` in the `decode()` function here: https://github.com/segmentio/kafka-go/blob/master/message.go#L65 

However, that's always `{nil}` for me. 

If I log the value of `c` [here](https://github.com/segmentio/kafka-go/blob/master/message.go#L67), it's always `0` which is [none](https://github.com/segmentio/kafka-go/blob/master/compression.go#L36)

**Comment by dkoston** [comment]
I tried kafka 0.11.0.1 as well and I'm still not able to see compressed messages.

Does the command line client also auto-decompress?

```
bash-4.3# kafka-console-consumer.sh --zookeeper zookeeper:2181 --topic topic-A --from-beginning
Using the ConsoleConsumer with old consumer is deprecated and will be removed in a future major release. Consider using the new consumer by passing [bootstrap-server] instead of [zookeeper].
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
long message that needs compression a a a a a a a a a a a a a a
```

**Comment by abraithwaite** [comment]
Hmm, you definitely should have those attributes set.

https://github.com/segmentio/kafka-go/blob/48c37f796910d8154479aaa04ade5843ae4f55d0/writer.go#L268-L271

Maybe there is something funky going on on the broker, but it's my understanding that all compression handling is done in the clients.

The command line client almost certainly decompresses though.

**Comment by dkoston** [comment]
Yea, I checked and the command line client definitely decompresses on-the-fly.


I was expecting that `message.CompressionCodec` would be set on messages coming out of the reader as it looks like it's set here: https://github.com/segmentio/kafka-go/blob/master/message.go#L73

However, by the time the code executes and I get a message, it's always `c = 0`.

I'll try to do some more digging but for my goals, it would be really nice to be able to see what compression codec is used. I know in newer versions of Kafka you can add headers which allows you to not have to Unmarshal() messages to check their type. I'm trying to implement the same by saying all message type A is code lz4 and all message type B s snappy. Should be faster to check the Attributes/`message.CompressionCodec` then to Unmarshal the protobuf and then check type and Unmarshal again into the final mesage format.

**Comment by dkoston** [comment]
Thanks for the quick responses as well BTW!

**Comment by abraithwaite** [comment]
Yeah, looking at your code and comparing it to the library code, nothing immediately stands out to us. 

We'll take a deeper look though.  My gut feeling is that it's reading the batch compressed, decompressing it, then not copying the attributes over to the message somewhere.

**Comment by dkoston** [comment]
Thanks both. As an FYI, my `00000000000000000000.log` file is growing at the exact same size when I remove the `CompressionCodec` option from the writer and write the messages as when the option is enabled. 

I sent messages with the first 4 chapters of "Alice in Wonderland" from here https://github.com/google/snappy/blob/master/testdata/alice29.txt to make sure they should be compressed however, each time, I sent 10 messages and the log file grew by 460211 bytes. I'd have expected the compressed messages to be smaller in the log.

**Comment by achille-roussel** [comment]
@dkoston we've done some work recently on the compression support in kafka-go, could you try out with a more recent version of the package and see if it's now working as you'd expect?

**Comment by stevevls** [comment]
I'm going to close this issue now.  I don't believe it to be an issue ever since #135.

**Comment by ajinkya-joshi-nutanix** [comment]
I am not able to produce messages using compression:
writer creator:
writer := kafka.Writer{
		Addr:       kafka.TCP(strings.Split((*model.KafkaBrokerURL), model.Comma)...),
		Topic:      producerAttributes.Topic,
		BatchBytes: *model.BatchBytesToProduce,
		BatchSize:  producerAttributes.BatchSize,
		Compression: kafka.Zstd
}
writer.writeMessages(context.Background(),
		kafka.Message{
			Key:   message.Key,
			Value: message.Value,
		},
	)
	
I dont see anything is getting produced. I am using aws MSK. Any. configuration is required?

**Comment by StartAgainJ** [comment]
> I'm going to close this issue now. I don't believe it to be an issue ever since #135.

@stevevls Still have this problem, my version is v0.4.21

**Comment by stevevls** [comment]
@ajinkya-joshi-nutanix @StartAgainJ I'm not an active contributor to this project anymore.  But whatever you're seeing, it's unlikely related to this issue because this was fixed in #135.  In this issue, messages were getting produced correctly but not compressed.  I would recommend you open a new issue so that the current maintainers see it.
